packages/dapi/lib/chainDataProvider/ChainDataProvider.js

```diff
@@ -0,0 +1,155 @@
+const { EventEmitter } = require('events');
+const { BlockHeader, ChainLock } = require('@dashevo/dashcore-lib');
+const log = require('../log');
+
+/**
+ * Data access layer with caching support
+ */
+class ChainDataProvider extends EventEmitter {
+  /**
+   *
+   * @param coreRpcClient {CoreRpcClient}
+   * @param zmqClient {ZmqClient}
+   * @param blockHeadersCache {BlockHeadersCache}
+   */
+  constructor(coreRpcClient, zmqClient, blockHeadersCache) {
+    super();
+
+    this.coreRpcAPI = coreRpcClient;
+    this.zmqClient = zmqClient;
+    this.blockHeadersCache = blockHeadersCache;
+
+    this.chainLock = null;
+  }
+
+  /**
+   * @private
+   * @param blockHash {string}
+   */
+  blockHashHandler(blockHash) {
+    this.emit(this.events.NEW_BLOCK_HEADER, blockHash);
+  }
+
+  /**
+   * @private
+   * @param rawChainLock {string}
+   */
+  chainLockHandler(rawChainLock) {
+    this.chainLock = new ChainLock(Buffer.from(rawChainLock, 'hex'));
+
+    this.emit(this.events.NEW_CHAIN_LOCK);
+  }
+
+  /**
+   * Grabs most recent chainlock
+   * @returns {Promise<void>}
+   */
+  async init() {
+    let chainLock;
+
+    try {
+      chainLock = await this.coreRpcAPI.getBestChainLock();
+    } catch (e) {
+      if (e.code === -32603) {
+        log.info('No chain lock available in dashcore node');
+      } else {
+        throw e;
+      }
+    }
+
+    this.chainLock = new ChainLock(chainLock);
```

Comment:
Shouldn't we emit `NEW_CHAIN_LOCK`? We probably just need to use `chainLockHandler` here.